File: ruby/98.validate-binary-search-tree.rb
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

# @param {TreeNode} root
# @return {Boolean}
def is_valid_bst(root)
  between_val?(root, -Float::INFINITY, Float::INFINITY)
end

def between_val?(node, min, max)
  return true if node.nil?
  return false if node.val <= min || node.val >= max

  between_val?(node.left, min, node.val) && between_val?(node.right, node.val, max)
end

File: rust/src/98.validate-binary-search-tree.rs
use crate::Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

#[allow(dead_code)]
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;

type Node = Rc<RefCell<TreeNode>>;

#[allow(dead_code)]
impl Solution {
    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        Self::is_between_val(root.as_ref(), None, None)
    }

    fn is_between_val(node: Option<&Node>, min: Option<i32>, max: Option<i32>) -> bool {
        if let Some(node) = node {
            let node = node.borrow();
            if min.map_or(false, |min_val| node.val <= min_val)
                || max.map_or(false, |max_val| node.val >= max_val)
            {
                return false;
            }

            Self::is_between_val(node.left.as_ref(), min, Some(node.val))
                && Self::is_between_val(node.right.as_ref(), Some(node.val), max)
        } else {
            true
        }
    }
}
